fhir2dataset/api_caller.py

```diff
@@ -232,3 +236,125 @@ def _init_data(self) -> dict:
+                    degres[j] = degres[j] - 1
+                else:
+                    degres[j] = degres[j] - 1
+                    somme = somme - 1
+        return (lst_cols, degres, somme)
+
+    def clean_join_columns(self):
+        """unnest join columns to allow the join 
+        - computes the degre (number of common nodes) between the join node(s) and all the other nodes :
+            1. Retrieves all join element of the current resource
+            2. Finds the tree containing the join element
+            3. Computes the degre between the join element and the other elements of the tree 
+        - for each dictionnary created ie each join element, calls explode(self.df,lst_cols) 
+        explode unnest the dataframe for columns nested to element join
+        """
+        degres_dict = dict()
+        cols_dict = dict()
+        elements = []
+        for element in self.elements.get_subset_elements(goal="join"):
+            dict_degre = self.compute_degres(element)
+            if dict_degre:
+                cols = [self.elements.get_element(i).col_name for i in dict_degre.keys()]
+                degres = [dict_degre[i] for i in dict_degre.keys()]
+                degres_dict[element.col_name] = degres
+                cols_dict[element.col_name] = cols
+                elements.append(element.col_name)
+
+        for i in range(len(elements)):
+            element = elements[i]
+            degres = degres_dict[element]
+            cols = cols_dict[element]
+            somme = 0
+            for degre in degres:
+                somme += degre
+            logger.info(f"the sum of degres is {somme}")
+            while somme > 0:
+                lst_cols, degres, somme = self.compute_lst_cols(degres, cols, somme)
+                logger.info(f"the list of columns nested to the join element are {lst_cols}")
+                if lst_cols:
+                    if i < len(elements) - 1:
+                        degres_dict = self.check_other_dict(
+                            elements, cols_dict, degres_dict, i, lst_cols
+                        )
+                    self.df = self.explode(self.df, lst_cols)
+                    somme = somme - len(lst_cols)
+
+    def explode(self, df, lst_cols):
+        # all columns except `lst_cols`
+        idx_cols = df.columns.difference(lst_cols)
+
+        lens = len(df[lst_cols[0]].iloc[0])
+
+        dic = self.to_list(df, lst_cols)
+
+        return (
+            pd.DataFrame({col: np.repeat(df[col].values, lens) for col in idx_cols})
+            .assign(**{col: dic[col] for col in lst_cols})
+            .loc[:, df.columns]
+        )
+
+    def to_list(self, df, lst_cols):
+        """explode the lists to rows 
+        """
+        dico = dict()
+        for col in lst_cols:
+            dico[col] = []
+        for index, row in df.iterrows():
+            for col in lst_cols:
+                for element in row[col]:
+                    dico[col].append(element)
```

Comment:
I think the whole function can be written
```
from collections import defaultdict

def to_list(self, df, lst_cols):
    dict = defaultdict(list)
    for _, row in df.iterrows():
        for col in lst_cols:
            dict[col].extend(row[col])
    return dict
```
But aren't you mixing the elements from the different rows here?